File: C++/number-of-islands.cpp
#include <bits/stdc++.h>
using namespace std;

void dfs(vector<vector<int>> &grid, int r, int c) {
    int rows = grid.size(), cols = grid[0].size();
    if (r < 0 || r >= rows || c < 0 || c >= cols || grid[r][c] != 1) return;
    grid[r][c] = 0; // mark visited
    dfs(grid, r + 1, c);
    dfs(grid, r - 1, c);
    dfs(grid, r, c + 1);
    dfs(grid, r, c - 1);
}

int main() {
    int rows, cols;
    cin >> rows >> cols;
    vector<vector<int>> grid(rows, vector<int>(cols));
    for (auto &row : grid) for (auto &c : row) cin >> c;

    int count = 0;
    for (int r = 0; r < rows; r++) {
        for (int c = 0; c < cols; c++) {
            if (grid[r][c] == 1) {
                count++;
                dfs(grid, r, c);
            }
        }
    }
    cout << count << endl;
    return 0;
}

File: JS/number-of-islands.js
function numIslands(grid) {
  const rows = grid.length, cols = grid[0].length;

  function dfs(r, c) {
    if (r < 0 || r >= rows || c < 0 || c >= cols || grid[r][c] !== 1) return;
    grid[r][c] = 0;
    dfs(r + 1, c);
    dfs(r - 1, c);
    dfs(r, c + 1);
    dfs(r, c - 1);
  }

  let count = 0;
  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      if (grid[r][c] === 1) {
        count++;
        dfs(r, c);
      }
    }
  }
  return count;
}

const lines = require('fs').readFileSync('/dev/stdin', 'utf8').trim().split('\n');
const [rows, cols] = lines[0].split(' ').map(Number);
const grid = [];
for (let i = 1; i <= rows; i++) grid.push(lines[i].trim().split(/\s+/).map(Number));

console.log(numIslands(grid));